src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -193,17 +307,24 @@ protected override void Free()
             BoundNode node,
             ArrayBuilder<(BoundReturnStatement, TypeSymbolWithAnnotations)> returnTypesOpt,
             VariableState initialState,
-            Action<BoundExpression, TypeSymbolWithAnnotations> callbackOpt)
+            PooledDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMap = null)
             : base(compilation, method, node, new EmptyStructTypeCache(compilation, dev12CompilerCompatibility: false), trackUnassignments: true)
         {
-            _callbackOpt = callbackOpt;
             _binder = compilation.GetBinderFactory(node.SyntaxTree).GetBinder(node.Syntax);
             Debug.Assert(!_binder.Conversions.IncludeNullability);
             _conversions = (Conversions)_binder.Conversions.WithNullability(true);
             _useMethodSignatureReturnType = (object)methodSignatureOpt != null && useMethodSignatureReturnType;
             _useMethodSignatureParameterTypes = (object)methodSignatureOpt != null && useMethodSignatureParameterTypes;
             _methodSignatureOpt = methodSignatureOpt;
             _returnTypesOpt = returnTypesOpt;
+
+            if (analyzedNullabilityMap == null)
+            {
+                analyzedNullabilityMap = PooledDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)>.GetInstance();
+                _ownsAnalyzedNullabilityMap = true;
+            }
+            _analyzedNullabilityMapOpt = analyzedNullabilityMap;
```

Comment:
>_analyzedNullabilityMapOpt [](start = 12, length = 26)

If you passed this map as a parameter, then this problem largely goes away. I think it's easy for each of the callers to determine whether to free the map or not, so the caller should be responsible for it.